### PR TITLE
Mobile alert notifications

### PR DESCRIPTION
**Pull request by mureytasroc**

### CODE REVIEW

File: backend/Pipfile
Comment (mureytasroc) [inline backend/Pipfile:21]: Make sure to undo this change before merging

File: backend/alert/views.py
Comment (mureytasroc) [inline backend/alert/views.py:92]: Why delete this?

File: backend/alert/views.py
Comment (mureytasroc) [inline backend/alert/views.py:101]: Why remove the else?

**Review by mureytasroc — COMMENTED** [review]
Generally looks good (I made a bunch of changes to make the whole close_notification thing more robust / fix all failing test cases; it turns out that is not a trivial feature to implement).  We definitely need some test cases though before we can merge (I can write some for the close_notification thing but you should def write some for push_notifications if possible).

File: backend/alert/alerts.py
Comment (davish) [inline backend/alert/alerts.py:72]: Did we decide on sending email close alerts? Not a cost issue, but thought we were keeping close alerts to mobile for now as an exclusive feature.

File: backend/alert/alerts.py
Comment (davish) [inline backend/alert/alerts.py:92]: Small thing, but if we want to make sure we're replying to the right thread, should probably pull out `%s is now open!` into its own string so we know it's the same.

File: backend/alert/alerts.py
Comment (davish) [inline backend/alert/alerts.py:95]: Should probably be using f-strings here and above too

File: backend/alert/alerts.py
Comment (davish) [inline backend/alert/alerts.py:157]: Should pull this out into an environment variable (see the bottom of `settings/base.py` for lots of examples

File: backend/alert/alerts.py
Comment (davish) [inline backend/alert/alerts.py:169]: Not sure if this needs to be in this PR, but we should probably hunt around for all of the hardcoded URLs and put them in a `constants.py` file or something

File: backend/alert/models.py
Comment (davish) [inline backend/alert/models.py:356]: if this function is the only place the above dict is used, should probably move this inline.

File: backend/alert/tasks.py
Comment (davish) [inline backend/alert/tasks.py:53]: Putting a default status here makes me kind of nervous, since this is data we should be relying on the webhook for. It might be a bit more work to set a variable at every call site, but I think it would be worth it to make sure bugs don't pop up.

File: backend/alert/tasks.py
Comment (davish) [inline backend/alert/tasks.py:57]: I think this is a really elegant solution to the problem! Is there any code that still uses `get_active_registrations` or can that code safely be deleted?

File: backend/alert/views.py
Comment (davish) [inline backend/alert/views.py:31]: Seconded here, about defaulting to "O" for the status.

File: backend/courses/models.py
Comment (davish) [inline backend/courses/models.py:450]: we love demorgan's law

File: backend/alert/alerts.py
Comment (davish) [inline backend/alert/alerts.py:122]: I think we'll need some more checking eventually to handle the case where users log out of penn mobile on their phones, how do we re-enable texts? but this can be a conversation with mobile at some point.

File: backend/tests/alert/test_alert.py
Comment (davish) [inline backend/tests/alert/test_alert.py:106]: some documentation on what this helper is doing would be useful!

File: backend/tests/alert/test_alert.py
Comment (davish) [inline backend/tests/alert/test_alert.py:158]: Again, docs on what the helper is doing would be great. Seems to be doing a lot that's hard to parse at first glance!

File: backend/tests/alert/test_alert.py
Comment (davish) [inline backend/tests/alert/test_alert.py:269]: Ditto again. Will stop commenting but yeah, docstrings on helper functions would be amazing!

File: backend/tests/alert/test_alert.py
Comment (davish) [inline backend/tests/alert/test_alert.py:1080]: Why are we ignoring updates to the push settings? seems like users should be able to change that from the client?

**Review by davish — CHANGES_REQUESTED** [review]
Really awesome work! Some small comments, and requests for a bit more documentation in parts. But should be good to go after those!

File: backend/alert/alerts.py
Comment (mureytasroc) [reply backend/alert/alerts.py:72]: So I was assuming standardization on the backend would be better and then we could regulate whether or not to afford the feature on the frontend (since close alerts are disabled by default, if the frontend doesn't change they won't be available to users on the web version).  Does this make sense or should I remove on the backend?

File: backend/alert/models.py
Comment (mureytasroc) [reply backend/alert/models.py:356]: It is also used in alert/tasks.py (line 47), to filter registrations (since you can't filter on a property):

`return list(section.registration_set.filter(**Registration.is_waiting_for_close_filter()))`

File: backend/alert/tasks.py
Comment (mureytasroc) [reply backend/alert/tasks.py:57]: Good point, it is no longer being used by anything.  Just deleted `get_active_registrations`

File: backend/tests/alert/test_alert.py
Comment (mureytasroc) [reply backend/tests/alert/test_alert.py:1080]: Sorry, I should really document these tests better.  This is testing the functionality that if the user only updates push_notifications, leaving out other fields, it ignores the other fields and only updates push_notifications.  I should really change the PCA PUT route to PATCH (have been meaning to do this)

File: backend/alert/alerts.py
Comment (mureytasroc) [reply backend/alert/alerts.py:169]: Agreed

File: backend/alert/tasks.py
Comment (mureytasroc) [reply backend/alert/tasks.py:53]: Fixed

File: backend/alert/alerts.py
Comment (mureytasroc) [reply backend/alert/alerts.py:122]: That's a good point; maybe we make enabling one automatically disable the other?

File: backend/alert/alerts.py
Comment (mureytasroc) [reply backend/alert/alerts.py:157]: Reached out to mobile backend to ask what the default bearer token should be... will move to an environment variable

File: backend/alert/alerts.py
Comment (r3khaan) [inline backend/alert/alerts.py:182]: Instead of returning `False` in line 182, shouldn't we be returning `None` since `self.registration.user` is `None` due to some error? (i.e. we do want to send a push notif, but there is an error preventing us from doing so)

File: backend/alert/templates/alert/push_notif.txt
Comment (r3khaan) [inline backend/alert/templates/alert/push_notif.txt:2]: I'm pretty sure that users can't click on individual links in mobile push notifications. Ideally, when we roll out the feature, the notification will have actions (accessed via the "...") like "go to pennintouch", "cancel alert", "resubscribe for alert", "manage alert in pennmobile".

We don't need to change it now, because I think it will remind us to add actions later, but just thought I'd give the heads up.

**Review by r3khaan — APPROVED** [review]
LGTM! Good stuff. Excited to roll this out soon.

File: backend/alert/alerts.py
Comment (mureytasroc) [reply backend/alert/alerts.py:182]: So this is actually behavior which is grandfathered in from the old PCA; self.registration.user is None on legacy registrations, so in that case it just wouldn't send any push notifications.  Not really an error state so I don't think returning None makes sense, but it's kind of an unreachable case now since there is no way of creating legacy registrations anymore.

File: backend/alert/alerts.py
Comment (mureytasroc) [reply backend/alert/alerts.py:122]: Maybe if push notifications are enabled we grey out the phone number box on the web frontend with some sort of button to disable push / enable text

Comment (astrike30) [reply backend/alert/alerts.py:121]: I think it might be worth emphasizing on the sign-up screen if alerts are going to the app or via text. The push notification token should still be good even if they get logged out of the app, and if they delete the app I think the next time they sign-up for a class they would know to change it in most cases?

File: backend/alert/templates/alert/push_notif.txt
Comment (astrike30) [reply backend/alert/templates/alert/push_notif.txt:2]: Not sure how this works on the mobile side, but does the notification go straight to pennintouch when tapped?

**Review by astrike30 — APPROVED** [review]
All looks good to me 👍  hopefully nothing breaks in production but I think features like this (sending notifications etc) are prone to being unstable at first. The docs are incredible btw!

File: backend/PennCourses/docs_settings.py
Comment (astrike30) [reply backend/PennCourses/docs_settings.py:338]: oh I did add a few comments above ^^ but I wasn't in review mode 🤦🏻‍♂️ just wanted to make sure everyone sees

Comment (mureytasroc) [reply backend/PennCourses/docs_settings.py:338]: So docs_settings.py actually is for the frontend docs.  Added your suggestions in the backend docs; check them out at `http://localhost:8000/admin/doc/models/alert.registration/` when the backend server is running.

File: backend/alert/alerts.py
Comment (davish) [reply backend/alert/alerts.py:74]: Should be fine to keep it in as long as we make sure to test that case

File: backend/alert/alerts.py
Comment (davish) [reply backend/alert/alerts.py:121]: Let's follow up on this with @ArmaanT and @r3khaan. Not sure how much of an issue it will be.

File: backend/alert/alerts.py
Comment (davish) [inline backend/alert/alerts.py:130]: We should never be sending close notifications via text messages -- this should be `return None` whether or not `close_text` is available.

**Review by davish — APPROVED** [review]
One pretty major comment (ALWAYS return `False` when trying to send a close notification for a text message, never even try to send) for spam and cost reasons. Otherwise, the other outstanding comment is how to handle the silent logout scenario on mobile. We need to know if we can still send push notifications if the user has happened to be logged out in penn mobile (because device ID is still stored? not sure how the API server works here) want to pull in @r3khaan and @ArmaanT to double-check

**Comment by r3khaan** [comment]
> One pretty major comment (ALWAYS return `False` when trying to send a close notification for a text message, never even try to send) for spam and cost reasons. Otherwise, the other outstanding comment is how to handle the silent logout scenario on mobile. We need to know if we can still send push notifications if the user has happened to be logged out in penn mobile (because device ID is still stored? not sure how the API server works here) want to pull in @r3khaan and @ArmaanT to double-check

We can send push notifs to users when they are logged out! The only case sending notifications fails is when the user deletes the app. Not sure what the protocol is for this case, but one idea is to send a request from mobile to backend confirming that the notif was received after every notification .. and if the request is not received, then we know to switch notif style from push notifs back to text notifs.

**Pull request merged**
